### ISSUE TITLE
Add `declaration-property-value-no-unknown`

### ISSUE THREAD

**Issue by caneta** [stylelint/stylelint#6376]
### What steps are needed to reproduce the bug?

```
body {
  padding: auto;
}
```

### What Stylelint configuration is needed to reproduce the bug?

```
{
  "extends": [
    "stylelint-config-standard",
    "stylelint-config-prettier"
  ],
  "rules": {
    "block-closing-brace-newline-after": [
      "always", {
        "ignoreAtRules": [ "if", "else" ]
      }
    ],
    "at-rule-no-unknown": [
      true,
      {
        "ignoreAtRules": [
          "mixin",
          "if",
          "else",
          "error",
          "include",
          "at-root",
          "extend"
        ]
      }
    ]
  },
  "property-no-unknown": true,
  "defaultSeverity": "warning"
}
```

### How did you run Stylelint?

From VSCode

### Which version of Stylelint are you using?

`14.13.0`

### What did you expect to happen?

The editor should tell me that 'auto' is not a permitted value for property 'padding'.

### What actually happened?

Nothing, the error is not reported.

### Does the bug relate to non-standard syntax?

Nope, tryed in a plain CSS

**Comment by ybiquitous** [comment]
@caneta Thanks for using the template.

Unfortunately, Stylelint doesn't support the validation of property-and-value pairs. So I recommend using the [`stylelint-csstree-validator`](https://www.npmjs.com/package/stylelint-csstree-validator) plugin. For example:

```console
$ cat package.json
{
  "devDependencies": {
    "stylelint": "^14.13.0",
    "stylelint-csstree-validator": "^2.0.0"
  },
  "stylelint": {
    "plugins": [
      "stylelint-csstree-validator"
    ],
    "rules": {
      "csstree/validator": true
    }
  }
}

$ cat test.css
body {
  padding: auto;
}

$ npx stylelint *.css

test.css
 2:12  ✖  Invalid value for "padding"  csstree/validator

1 problem (1 error, 0 warnings)
```

See also our [document about validators](https://stylelint.io/about/linting/#validators).

**Comment by jeddy3** [comment]
This question pops up often, and it feels like an appropriate time to revisit it.

As mentioned in [our doc](https://stylelint.io/about/linting/#validators)  that @ybiquitous linked to, there are three types of tools:
- linters
- validators
- formatters (aka pretty printers)

Stylelint has historically overlapped with both validators and formatters. We intend [to remove our formatting rules](https://stylelint.io/user-guide/rules/list#stylistic-issues) so that we don't overlap with formatters. We need a clear vision for how we approach validation too.

We split [our rules](https://stylelint.io/user-guide/rules/list/) into two groups that:

- [avoid errors](https://stylelint.io/user-guide/rules/list#avoid-errors)
- [enforce (non-stylistic) conventions](https://stylelint.io/user-guide/rules/list#enforce-conventions)

The latter group is made up of 5 subgroups that:

- disallow CSS features with the `no-*`, `*-allowed-list` and `*-disallowed-list` rules , e.g. [`unit-disallowed-list`](https://stylelint.io/user-guide/rules/list/unit-disallowed-list)
- apply limits with the `max-*` and `min-*` rules, e.g. [`selector-max-id`](https://stylelint.io/user-guide/rules/list/selector-max-id)
- enforce naming conventions with the `*-pattern` rules, e.g. [`custom-property-pattern`](https://stylelint.io/user-guide/rules/list/custom-property-pattern)
- enforce one representation of things that have multiple with the `*-notation` rules, e.g. [`color-function-notation`](https://stylelint.io/user-guide/rules/list/color-function-notation)
- require or disallow quotes with the `*-quotes` rules, e.g. [`function-url-quotes`](https://stylelint.io/user-guide/rules/list/function-url-quotes)
- disallow redundancy with the `no-redundant-*` rules, e.g. [`shorthand-property-no-redundant-values`](https://stylelint.io/user-guide/rules/list/shorthand-property-no-redundant-values)

These all apply to valid CSS. They are powerful rules, mostly unique to Stylelint and aligned with our goal of being a linter.

The group that avoids errors is made up of two subgroups that check for:

- invalid CSS, e.g. rules that disallow:
   - malformed things with the `*-no-invalid` rules , e.g. [`named-grid-areas-no-invalid`](https://stylelint.io/user-guide/rules/list/named-grid-areas-no-invalid)
   - unknown features with the `*-no-unknown` rules, e.g. [`unit-no-unknown`](https://stylelint.io/user-guide/rules/list/unit-no-unknown)
- valid CSS that may have unintended consequences, e.g. rules that disallow:
   -   duplicates with the `no-duplicates-*` rules, e.g. [`keyframe-block-no-duplicate-selectors`](https://stylelint.io/user-guide/rules/list/keyframe-block-no-duplicate-selectors) 
   - overrides with the `no-*-overrides` rules, e.g. [`declaration-block-no-shorthand-property-overrides`](https://stylelint.io/user-guide/rules/list/declaration-block-no-shorthand-property-overrides)
   - missing things with the `no-missing-*` rules, e.g.  [`custom-property-no-missing-var-function`](https://stylelint.io/user-guide/rules/list/custom-property-no-missing-var-function)
   - ignored things with the remaining `no-*` rules, e.g. [`keyframe-declaration-no-important`](https://stylelint.io/user-guide/rules/list/keyframe-declaration-no-important) (as an aside, we should have called [`no-invalid-double-slash-comments`](https://stylelint.io/user-guide/rules/list/no-invalid-double-slash-comments) `no-double-slashes` as they're neither invalid nor comments.)

The former group overlaps with validators whereas the latter does not. Like the rules that [enforce (non-stylistic) conventions](https://stylelint.io/user-guide/rules/list#enforce-conventions), the ones in the latter subgroup are mostly unique to Stylelint.

Additionally, we catch and surfaces broad syntax errors, e.g. [unclosed blocks](https://stylelint.io/demo#N4Igxg9gJgpiBcICGACUAaEBnAngOwBckAPBcLLETSPAMwEsBzM4AHTxRVZACcBXADYws3eGnadO3JHjwQiBehDwBaOSr54A1nIDueUSgL8Y6CZOkEV-IWogbtegwiMmzHCyABGAiGC12KjAAtgAOBDiGxnym5lLgEL48gfR4AG5IAvRQKgAWMMRRbnFcCcHBMISBIeGRLtGxHvFgfFgEEMEqoTwQoTA8EYHB9BSpjCoZybSaYIrKRTHukqWwYAJIPEhzqj5+AepQfKFZYFswKi1tHV09fQP0wguNy9yr65vbKrv+gYfH9KcCOdur1+opHi4ANolTgNJbLcRNBHcJhyHgwQzQpEI0o0LAwFqKNLnP4nM5YFS6egEXIqKD0Wi0fqVKwZAQxEQgGGSAC63IAviUefD4m8NlslDtfD91FhchABrkZDkQXdBhBiTweNkIWI4SVuLRlFZaEhhgIcL8jmSgWozbrXIsDSAjVVTebLephqM8ONGJV+gCVO76BaVFoYDhdAqoE8RaVpnhZpKLpkwIFNFhQkgwDAcqCPgq487E8nlCosngYBsVIxNvSWYE5Hg2sqNjl6eiy849cUkYaZp91JodBB9MX+yAIzhaJsKl9pft7KSAWcVPihLMFZze07J9PZ-a6QT3hLy+p6GEFURCBOXiAKvSkMHqwQ+Oi7fPh44xz3Hc9PDwe0cjrbIVA2asKQvdJMmyO9AOXYRczwelfXXPowAZAFqTqXcAPiA5rVXW0thUS9QmvaxBAdfVJ0I-5AXODcCXaHgd3-eNuHUGpBiwCB31zeCCPsApjGffFhkgXwWyE0poLZMCoH4nwmPWOULg6CpCHY2j73k2CVQgLBqRTUjyMomwMXqPs9PsegtRgRhBBrXRcmpYRs0E6y93vVUwU9ewRycWTuGYrdklCfE+CUi41KgwKf3Hbz8NKMLWK6KKYpgIQtKsb9RySvDOOwbKWIVFQIj6DNEr-LEcV0hE2GxTxUQVKyxDqnFPEudpOlK3LOW5Tg+WawUkWFZ02m1ND1CrXRK3ajjnU0alqoKv8GviTJQiVCZMhiOwFElTFuW4O5kKIf1uHjTgmq60oClzcIAAVbjBB52M6+7uF6HMcOuobPAYAQBBUX7MIiAHmvvWhfGgMHPP+qhAfiK5QgRv7IeR6HPCmiAIwxiG6kBkacTG5YJsnUjLKCMJBgWr4YCNdETua6QBF0JAcE5G7EW+kBHpgcJWfu+Jvi0IQKHA2ggWSLAjyA+dxclnmUYTey2jUYQgVjLkccp-nWpZqE1ekGX+g08oWW4EmBSFYrqeoz9zlOfFDG4XxdH6KH70d2xFaYzzziQc2eHd5AOa5iljN9WwFp9zw-fOdRiRQ8ruiZ+hCmSh2rBpiSAUSc8YHm1Jg9D8PMk57mE7FxdYqMsYvk2XNadqCsy8Z5nFq4mBNVr0pxYbmPxi8Fvk5LhmQ9lyvI5r7H7yHtZG7Qsecwn0uqy7trZ+rilgkERQO6rAfuCX3wR+b9f0OvrwmZ3lx2b39cxjjsvT+8evQTwJu19buap4V0fhHZ+B8BBH3jgvTwQ9v6-3HjfVu09vbAKrlHF+sdziQOKjAvoP9V7wKzLfe+xsxBPyjh-KS5VSxDnkGeP83BgjQH6M4YqlDkj5GIKmN2wDPbINYUXdhBQKyVEYDScOcprwUM0o2HiloGZ327iLZEIDyFQOWHdLq3BBbCxNjjeIDA2J5W1nmG2+tebxCNotL6mjsARCEJWKwkByjKkGvrO241+FWyqK5dyhDW6pGMrAXeqjWGtF6jcAsgxZHH3OAoh+HVToqPnrzDROItHECegQJR-MkHJB6tcPyAw6jmKBhrIxbQTF63ugbGxljsk2NyZbXKA80kgACTqdBjA35b3FqY6p7iKahKuJ0R89BnzZgILLWqiSAB6AAKSEz4ABePJFkqCWQABhUAATh5AAKgAJRzJUGszZOyeQAGoDl7IACQtL5jYioFAkBXWAQAUWIBhHWKB8nBBQKM1AABHGIPAcAoADkYCAKA74oAjF4JAXguFWQGbyIZ4TCmDAmVM+p8R5mnNWcsrZuzDnHNOUSy51y7lqMkKk5RTz5avNISAD5Xy8w-LCR0FAGKwUQvaNCmAsKYDwsRa7ZFo17bOjFB8FM8K0J+PLjPYBVZ+7UteCecUnxZXjHldvEhpRUHJMleq6V5Yh4F0oaoABndcnBO5ioMBED36qpAFKuhC49jrhCIXaSCCFXIKZQa6Or9MFOuKq6zV9dzVF1UDquJere4qrDcat1Zrg0xPtSQY8awNWSnYgARiTdmk1UoPViVDE3KN0lbWq0nOGmV9d1DokOChGQVhpKMCVChCJaoPohRdcmz4Fqtabz9WHFBc996H3oDEj+dbyxDp1Ta8dz8R7dKsoW08g7o2+t1T3EAyq+FGqLW66J8jiFWMSYG+5tL7zaKybo0W+rQ5Zs3cdalrSDGayrBU3WJMSkokYGiC9ein2yyadbd9yj2mwE6Wu91-g+ldVJgicmqKSzGmDGaUMnojzAvkA6MhdrXLMmsASaRKFKnFWoSmJxwRnxWq3kugNE77VTpnc66j87NLiSDtLRVzGV1pqwSWQcNHuM7rjXug9Y6qOifLHRzhp6y7sQ2bJpMQ4jyivDrwmTIn1MpmzOiQg+R8RQUnp3aDe7A2sfAdO4Tk5OOqEMyyEzwgd2WfDtJ2DIaT4cbk6od8oM8NAnYoRmtMN-OUjciF3jTH9UTo-rkA6sA6wwGTrQ7YlcMEf3MgMQ6dDxHGDGDlijhACsuAAEzFQPHOJipVwr5cy8A86LIXnnE0CrSk1I3KqGnNGHg+Y8Bhl6c6mr9ooJHixcwnFpQ8XLIJes8lJKTmEvOVc2517AYMOEAyvdLKWJsrGxUcF9pIX8sFcKpFiHlioeGsVIQvoaTrP6PYb81I+0KbbvTZThgC3Ok+wzB7ojci-Yq6p-7eYxkvi2O+F227F1AIEyEiHT5odvg-Au3jknPN90PZOAFaPYfOyuzwsceP7wE6ZDDj8AdAipyUhFdEDBs5FRR1Dqn6PgQQWM8INzOqPNKtx7p-HkPnwc6J5sX05wCxbHKgj-j8W94f0p6+CXMh-QI36LLuWWPz3VuV6LlQwL+hyJGI48TDHR3Vps463zxUCfG9BR3TWtGePXzi2F7z7H7eG8d6bl34nY168F4m503E6aWk-W2p1rO6IiUSFF3xQc+1ehGJfSo+ZaDrn4jwf+5nfOx70nwYId9khCCQKhcYSyXv6+dXgYvpeM2cIzphYy8wXAABZir15LxbdQZbKxV5ezpGynhuUk6ZTpj+4+U4Z-TkzrOfaaZKZ6cHhJbMkmTts977kN7PB3pmzDMpWsf3XYRDU1pdSH2izNmB2jEHbbio8ZNer6UthFa8HwW0f8IwEGjrxgXJlaTD+NKcqd-bUT-W0GXdKeXf1OSIXEA1-MAyZCAr-aXO4bXCTNfeA0PScUA5IcA+gSA84YLAjTfRAzcdKbNKWKbHgaZDfObdZBbM5YlI5FbRbNbSlTbHGbbZ5RlUofbWYNlagrAFAfAs7GFOFBFEnFFO7F-Sg8qJxLwVITA2AsdJHQ1PApAvJDoZQoCGA3XRRZdZHLQhQ5IWALAZCKAVtS2PQzA2acseVPtfAsiFUd-aba-e8RglZMlc5ZbXw3ZdbKlFJLbB8HbNrcOQQ75bIMQ7QiQgVKQkVJAbhWQlAC-VKbQ53c3ZxYdQBBXMLCgsqMvM3JpN3VubHEPcnXGTIz7FfB0cHUwoounOfRnTOFnJaRohrOQPKDLN9JlJxY4AoQohrSKGAaKewEQifUoKfZ1Fw0Y8Y2KZIikZzHnUzdzFsHUHHXA+8OYzKESHKRsLTUnL2YXHYzI+YrKA4qoIdbo8rJlJSPgFSYY9KSqF2ZIvdGY4qCRRUZUbtfyQIJtTQawqoNkDkZworNCUg0LF1ZSIQD+FaRxd47TMnU4zwUE84PrGMKYj2FEj+dErI0o3I61RHRXNBB1OzUNZ0fErIAPHItQ63VdHzddKk-aENWkujLAowoAhA51akko2oiPGJFTYqfE2fNOVo5nJ4cwMafkEAfkIAA).

The complete list of our validating rules is:

- [`annotation-no-unknown`](https://stylelint.io/user-guide/rules/list/annotation-no-unknown)
- [`at-rule-no-unknown`](https://stylelint.io/user-guide/rules/list/at-rule-no-unknown)
- [`function-no-unknown`](https://stylelint.io/user-guide/rules/list/function-no-unknown)
- [`selector-pseudo-class-no-unknown`](https://stylelint.io/user-guide/rules/list/selector-pseudo-class-no-unknown)
- [`selector-pseudo-element-no-unknown`](https://stylelint.io/user-guide/rules/list/selector-pseudo-element-no-unknown)
- [`selector-type-no-unknown`](https://stylelint.io/user-guide/rules/list/selector-type-no-unknown)
- [`media-feature-name-no-unknown`](https://stylelint.io/user-guide/rules/list/media-feature-name-no-unknown)
- [`unit-no-unknown`](https://stylelint.io/user-guide/rules/list/unit-no-unknown)
- [`color-no-invalid-hex`](https://stylelint.io/user-guide/rules/list/color-no-invalid-hex)
- [`named-grid-areas-no-invalid`](https://stylelint.io/user-guide/rules/list/named-grid-areas-no-invalid)
- [`no-invalid-position-at-import-rule`](https://stylelint.io/user-guide/rules/list/no-invalid-position-at-import-rule)
- [`string-no-newline`](https://stylelint.io/user-guide/rules/list/string-no-newline)

Of which, the following overlap with CSSTree:

- [`at-rule-no-unknown`](https://stylelint.io/user-guide/rules/list/at-rule-no-unknown)
- [`function-no-unknown`](https://stylelint.io/user-guide/rules/list/function-no-unknown)
- [`unit-no-unknown`](https://stylelint.io/user-guide/rules/list/unit-no-unknown)
- [`color-no-invalid-hex`](https://stylelint.io/user-guide/rules/list/color-no-invalid-hex)
- [`string-no-newline`](https://stylelint.io/user-guide/rules/list/string-no-newline)

Now that we have the lay of the land, let's decide on a clear vision for validation.

I think we have three options:

- keep our current approach of adding validation for simple cases
- remove our validation rules, i.e. the `*-no-unknown` and `*-no-invalid` ones
- add built-in rules to validate pairings like property/value and at-rule/params ones

For the first two, we'd need to improve our documentation to better explain linters and validators, and how Stylelint can be used with a validator like CSSTree.

There are pros and cons to all three approaches, for example:

- our current approach muddies the water, i.e. we offer some validation but not for pairings like property/value ones and this confuses users
- removing our validation rules would be clear-cut for our uses, but current validators don't catch things like invalid named grid areas
- to add built-in rules we may need to bundle a 2nd parser, e.g. CSSTree, although we've spoken about adopting specific parsers for tricky rules in https://github.com/stylelint/stylelint/issues/5586

Based on how often people open property/value pair validation issues in our tracker, I feel users are expecting this feature. Especially as some CSS-in-JS libraries like [vanilla-extract](https://github.com/vanilla-extract-css/vanilla-extract) now support this (via the [CSSType library](https://github.com/frenic/csstype)). 

What do people think?

It's a big topic and one that may refine the direction of Stylelint. So let's leave this issue open as a discussion for a while so that we can stew on it.

**Comment by caneta** [comment]
> @caneta Thanks for using the template.
> 
> Unfortunately, Stylelint doesn't support the validation of property-and-value pairs. So I recommend using the [`stylelint-csstree-validator`](https://www.npmjs.com/package/stylelint-csstree-validator) plugin. For example:
> 
> ```
> $ cat package.json
> {
>   "devDependencies": {
>     "stylelint": "^14.13.0",
>     "stylelint-csstree-validator": "^2.0.0"
>   },
>   "stylelint": {
>     "plugins": [
>       "stylelint-csstree-validator"
>     ],
>     "rules": {
>       "csstree/validator": true
>     }
>   }
> }
> 
> $ cat test.css
> body {
>   padding: auto;
> }
> 
> $ npx stylelint *.css
> 
> test.css
>  2:12  ✖  Invalid value for "padding"  csstree/validator
> 
> 1 problem (1 error, 0 warnings)
> ```
> 
> See also our [document about validators](https://stylelint.io/about/linting/#validators).

Thank you, I have used CSSTree indeed and everything worked as expected.
@jeddy3 is right, I was expecting that writing something like `padding: inventedValue` was notified by the linter, so it could be useful to go ahead with this topic.

**Comment by ybiquitous** [comment]
@jeddy3 Thanks for the understandable classification of the built-in rules. It might be wonderful if we could reflect this grouping on our documentation somehow.

In addition, thanks for providing several future policies for validation rules.

I believe many people would like to receive the benefits of validation rules out of the box. So, now I'm inclined to the third option:

> add built-in rules to validate pairings like property/value and at-rule/params ones

although there may be some problems like a larger bundle size or higher maintenance costs.

**Comment by jeddy3** [comment]
> I believe many people would like to receive the benefits of validation rules out of the box. So, now I'm inclined to the third option.

SGTM.

> although there may be some problems like a larger bundle size or higher maintenance costs.

I agree. It feels like the pros outweigh the cons, though. Unless anyone knows of an alternative, we can use the CSSTree within the rule for now. We can refactor the implementation if a more appropriate library comes along (or if we end up rolling our own).

I believe the CSSTree parser returns everything we need, including what we need for accurate positions (i.e. `mismatchOffset` and `mismatchLength`):

```js
import { lexer } from "css-tree";
lexer.matchProperty("padding", "10px auto");
```

```console
{
  matched: null,
  iterations: 24,
  error: SyntaxError [SyntaxMatchError]: Mismatch
      at matchSyntax (file:///Users/jeddy3/Projects/temp/node_modules/css-tree/lib/lexer/Lexer.js:82:17)
      at Lexer.matchProperty (file:///Users/jeddy3/Projects/temp/node_modules/css-tree/lib/lexer/Lexer.js:324:16)
      at file:///Users/jeddy3/Projects/temp/csstree.mjs:7:21
      at ModuleJob.run (node:internal/modules/esm/module_job:193:25)
      at async Promise.all (index 0)
      at async ESMLoader.import (node:internal/modules/esm/loader:533:24)
      at async loadESM (node:internal/process/esm_loader:91:5)
      at async handleMainPromise (node:internal/modules/run_main:65:12) {
    message: 'Mismatch\n' +
      '  syntax: [ <length> | <percentage> ]{1,4}\n' +
      '   value: 10px auto\n' +
      '  -------------^',
    rawMessage: 'Mismatch',
    syntax: '[ <length> | <percentage> ]{1,4}',
    css: '10px auto',
    mismatchOffset: 5,
    mismatchLength: 4,
    offset: 5,
    line: 1,
    column: 6,
    loc: { source: '<unknown>', start: [Object], end: [Object] }
  },
  getTrace: [Function: getTrace],
  isKeyword: [Function: isKeyword],
  isProperty: [Function: isProperty],
  isType: [Function: isType]
}
```

We'll want to limit the scope of the rule [compared to the plugin](https://github.com/csstree/stylelint-validator/blob/master/lib/index.js), e.g. the rule should only check standard CSS for unknown values within properties. We should silently catch all other errors the parser throws, including the error for unknown properties so that the rule [doesn't overlap](https://stylelint.io/user-guide/rules/) with [our existing `property-no-unknown` one](https://stylelint.io/user-guide/rules/property-no-unknown). 

We may also want to limit our [`unit-no-unknown`](https://stylelint.io/user-guide/rules/unit-no-unknown) and [`function-no-unknown`](https://stylelint.io/user-guide/rules/function-no-unknown) rules to at-rules params so that they won't overlap with this new rule.

- Name: `declaration-property-value-no-unknown`
- Primary option: `true`
- Secondary options:
   - `ignoreProperties: []`
   - `ignoreValues: []`
   - `propertiesSyntax: {}`
   - `typesSyntax: {}`
- Autofixable: No
- Message: `Unexpected unknown value "${value}" for property "${property}"`
- Description: "Disallow unknown values for property within declarations"
- Extended description: "This rule considers values for properties defined within the CSS specifications to be known. You can use the `propertiesSyntax` and `typesSyntax` secondary options to alter the syntax."
- Section: "Avoid errors" -> "Unknown"

I've labelled the issue as ready to implement. Please consider [contributing](https://stylelint.io/contributing) if anyone fancies giving this ago.

There are [steps on how to add a new rule](https://stylelint.io/developer-guide/rules#add-a-rule) in the Developer guide.
